include/dlaf/eigensolver/bt_band_to_tridiag/impl.h

```diff
@@ -328,6 +351,61 @@ struct TileAccessHelper {
   Part part_bottom_;
 };
 
+struct DistIndexing {
+  DistIndexing(const TileAccessHelper& helper, const matrix::Distribution& dist_hh, const SizeType b,
+               const GlobalTileIndex& ij, const SizeType& ij_b_row)
+      : dist_hh(dist_hh), b(b), mb(dist_hh.blockSize().rows()), helper(helper), ij(ij),
+        ij_b_row(ij_b_row) {
+    rank = dist_hh.rankIndex();
+    rankHH = dist_hh.rankGlobalTile(ij);
+    n_ws_per_block = to_SizeType(static_cast<size_t>(std::ceil(mb / b / 2.0f)) + 1);
+  }
+
+  comm::IndexT_MPI rankRowPartner() const {
+    return (rankHH.row() + 1) % dist_hh.commGridSize().rows();
+  }
+
+  bool isInvolved() const {
+    const bool isSameRow = rank.row() == rankHH.row();
+    const bool isPartnerRow = rank.row() == rankRowPartner();
+    return isSameRow ||
+           (dist_hh.commGridSize().rows() > 1 && isPartnerRow && helper.affectsMultipleTiles());
+  }
+
+  LocalTileIndex wsIndexHH() const {
+    const SizeType row = [&]() -> SizeType {
+      if (rank.row() == rankHH.row()) {
+        // Note: index starts at 1 (0 is the extra workspace), moreover max half blocks will run in parallel
+        const SizeType intra_idx = 1 + (ij_b_row % (mb / b)) / 2;
+        DLAF_ASSERT_HEAVY(intra_idx < n_ws_per_block, intra_idx, n_ws_per_block);
+        return dist_hh.localTileFromGlobalTile<Coord::Row>(ij.row()) * n_ws_per_block + intra_idx;
+      }
+      else {
+        DLAF_ASSERT_HEAVY(helper.affectsMultipleTiles() && (rank.row() == rankRowPartner()),
+                          helper.affectsMultipleTiles(), rank.row(), rankRowPartner());
+        return dist_hh.localNrTiles().isEmpty()
+                   ? 0
+                   : dist_hh.localTileFromGlobalTile<Coord::Row>(ij.row() + 1) * n_ws_per_block;
+      }
+    }();
+    return {row, 0};
+  }
+
+protected:
```

Comment:
Style question: do you have a reason to prefer `protected` over `private`? Not a blocker.

Reply:
There is no real reason, in this case it is just a semi-random way to remove those from the public API.

About these classes, I'm almost convinced that they should be declared `final`...but for the moment I would let it go like they are.

Reply:
Thanks. Yeah, `final` also sounds good. Just something to keep in mind for future changes. I'll mark this as resolved as my question was answered.